qbraid_qir/qasm3/visitor.py

```diff
@@ -502,7 +518,24 @@ def _transform_gate_params(self, gate_op: QuantumGate, param_map: dict) -> None:
                 gate_op.arguments[i] = param_map[param.name]
             # TODO : update the arg value in expressions not just SINGLE identifiers
 
-    def _visit_custom_gate_operation(self, operation: QuantumGate) -> None:
+    def _validate_gate_call(
+        self, operation: QuantumGate, gate_definition: QuantumGateDefinition, qubits_in_op
+    ) -> None:
+        if len(operation.arguments) != len(gate_definition.arguments):
+            self._print_err_location(operation.span)
+            raise Qasm3ConversionError(
+                f"""Parameter count mismatch for gate {operation.name.name}. Expected \
+{len(gate_definition.arguments)} but got {len(operation.arguments)} in operation"""
+            )
+
+        if qubits_in_op != len(gate_definition.qubits):
+            self._print_err_location(operation.span)
+            raise Qasm3ConversionError(
+                f"""Qubit count mismatch for gate {operation.name.name}. Expected \
+{len(gate_definition.qubits)} but got {qubits_in_op} in operation"""
+            )
+
+    def _visit_custom_gate_operation(self, operation: QuantumGate, inverse: bool = False) -> None:
```

Comment:
Same note as above, if we could more explicitly define the effect of inverse = True/False here and if applicable, add some type of validation to account for operations for which inverse = True is not valid.